### PR TITLE
feat: install pypi dependencies (alpha)

### PR DESCRIPTION
**Pull request by baszalmstra**
This is an initial version to support pypi dependencies (#175).

There is still lots of functionality missing but the main implementation works. Just add:

```toml
[pypi-dependencies]
plotly = ">=5.18.0" # or whatever
```

to your `pixi.toml` and enjoy.

This work builds on https://github.com/prefix-dev/rip which means `sdists` are currently not (yet) supported.

### CODE REVIEW

File: src/project/python.rs [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+}
+
+impl PythonDependencies {
+    /// Returns `true` if no requirements have been specified
+    pub fn is_empty(&self) -> bool {
+        self.requirements.is_empty()
+    }
+
+    /// Returns the requirements as [`pep508_rs::Requirement`]s.
+    pub fn as_pep508(&self) -> Vec<pep508_rs::Requirement> {
```
Comment (ruben-arts) [inline src/project/python.rs:25]: Looking at https://peps.python.org/pep-0508/ `"*"` should be a valid version but i get the following error:
```
 × failed to parse pixi.toml from /home/rarts/testpy
  ╰─▶ failed to parse project manifest
    ╭─[pixi.toml:15:1]
 15 │ [python-dependencies]
 16 │ pytest = "*"
    ·          ─┬─
    ·           ╰── data did not match any variant of untagged enum RequirementModern
    ╰────
    ```

File: src/project/manifest.rs [hunk 2]
```diff
@@ -66,6 +67,10 @@ pub struct ProjectManifest {
     /// We use an [`IndexMap`] to preserve the order in which the items where defined in the
     /// manifest.
     pub activation: Option<Activation>,
+
+    /// Optional python requirements
+    #[serde(default, rename = "python-dependencies")]
+    pub python_dependencies: PythonDependencies,
```
Comment (ruben-arts) [inline src/project/manifest.rs:73]: As talked about in discord lets make this `pypi-dependencies`

File: src/lock_file/mod.rs [hunk 3]
```diff
@@ -0,0 +1,521 @@
...
+    .await
+    .unwrap_or_else(|e| Err(e).into_diagnostic())
+}
+
+/// Returns true if the locked packages match the dependencies in the project.
+pub fn lock_file_up_to_date(project: &Project, lock_file: &CondaLock) -> miette::Result<bool> {
+    let platforms = project.platforms();
+
+    // TODO: Add support for python dependencies
+    if !project.python_dependencies().is_empty() {
```
Comment (ruben-arts) [inline src/lock_file/mod.rs:50]: We should also check here if there are pip managed dependencies in the lockfile to regenerate for removal of any dependencies. 

As going from x to 0 python dependencies doesn't trigger the update.

**Review by ruben-arts — COMMENTED** [review]
We can discuss what we change before the first merge.

File: src/environment.rs [hunk 4]
```diff
@@ -186,620 +187,468 @@ pub async fn update_prefix(
...
-    // If this fails, the error will be propagated using `?`
-    let Ok(channel) = Channel::from_str(channel, &Default::default()) else {
-        return false;
-    };
-
-    // Check if the URL starts with the channel's base URL
-    // Return true or false accordingly
-    url.starts_with(channel.base_url.as_str())
+#[derive(Debug, Clone)]
+struct MessageFormatter {
```
Comment (baszalmstra) [inline src/environment.rs:432]: Comments plox!

File: src/project/python.rs (hunk 1 above)
Comment (baszalmstra) [reply src/project/python.rs:25]: Lets also do in another PR.

File: src/lock_file/mod.rs (hunk 3 above)
Comment (baszalmstra) [reply src/lock_file/mod.rs:50]: Lets do in another PR when we integrate the lock-file better

File: src/lock_file/python.rs [hunk 5]
```diff
@@ -0,0 +1,370 @@
...
+        }
+    } else if platform.is_linux() {
+        let max_glibc_version = match &project.manifest.system_requirements.libc {
+            Some(LibCSystemRequirement::GlibC(v)) => v.clone(),
+            Some(LibCSystemRequirement::OtherFamily(_)) => return Vec::new(),
+            None => default_glibc_version(),
+        };
+        linux_platform_tags(platform, &max_glibc_version)
+    } else {
+        todo!("no implementation for mac yet!")
```
[the remark this answers is not in the document]
Comment (baszalmstra) [reply src/lock_file/python.rs:183]: See https://github.com/prefix-dev/pixi/issues/465

File: src/project/python.rs (hunk 1 above)
Comment (baszalmstra) [reply src/project/python.rs:25]: See https://github.com/prefix-dev/pixi/issues/466

File: src/lock_file/mod.rs (hunk 3 above)
Comment (baszalmstra) [reply src/lock_file/mod.rs:50]: Done! https://github.com/prefix-dev/pixi/issues/467

**Review by ruben-arts — APPROVED** [review]

**Pull request merged**